src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq_vmr.c

```diff
@@ -3019,43 +3021,109 @@ static ssize_t program_vmr_store(struct device *dev,
 }
 static DEVICE_ATTR_WO(program_vmr);
 
-static ssize_t vmr_system_dtb_read(struct file *filp, struct kobject *kobj,
-	struct bin_attribute *attr, char *buf, loff_t off, size_t count)
+static ssize_t vmr_default_fpt_read(struct file *filp, struct kobject *kobj,
+        struct bin_attribute *attr, char *buf, loff_t off, size_t count)
 {
-	struct xocl_xgq_vmr *xgq =
-		dev_get_drvdata(container_of(kobj, struct device, kobj));
-	char *blob = NULL;
-	size_t size = 0;
-	ssize_t ret = 0;
+        struct xocl_xgq_vmr *xgq =
+                dev_get_drvdata(container_of(kobj, struct device, kobj));
+        char *log_buf = NULL;
+        size_t log_size = 0;
+        ssize_t ret = 0;
 
-	/*
-	 * refresh cached data only if off is 0,
-	 * which indicates that there is a new request.
-	 */
-	if (off == 0)
-		ret = xgq_refresh_system_dtb(xgq);
+        /* return count should be less or equal to count */
+        ret = xgq_vmr_default_fpt(xgq, &log_buf, &log_size, off, count);
+        if (ret)
+                return (ret == -EINVAL) ? 0 : ret;;
 
-	if (ret)
-		return ret;
+        if (off  >= log_size)
+                goto done;
 
-	mutex_lock(&xgq->xgq_lock);
+        if (off + count > log_size)
+                count = log_size - off;
 
-	blob = xgq->xgq_vmr_system_dtb;
-	size = xgq->xgq_vmr_system_dtb_size;
+        memcpy(buf, log_buf + off, count);
+        vfree(log_buf);
```

Comment:
can you please move the vfree after done:?
otherwise, we have memory leak when goto done.